### PR TITLE
feat: produce metadata in optimizer output

### LINKED ISSUE #65: Produce metadata along with RelNode
- add group id metadata so DAG can be identified
- remove physical collector from `optd-core`

### PR DESCRIPTION
**Pull request by yliang412**
Closes #65. This PR produces metadata in optimizer output. Currently, the only metadata kept is the group id. The group id information has the following benefits:
- allow the consumer of the optd plan to recover reusable fragments in the DAG-shaped query plan.
- implement adaptive physical collector outside `optd-core`.
- display cardinality cost: #89 

Looking forward to feedbacks!

In the example below, we produce the same plan as before when adaptiveness is enabled.

```sql
❯ create table t1(t1v1 int, t1v2 int);
0 rows in set. Query took 0.009 seconds.

Execution took 0.000 secs, Planning took 0.000 secs
❯ explain select * from t1 as a, t1 as b;
+--------------------------------------------------+-------------------------------------------------------------------------------------------------+
| plan_type                                        | plan                                                                                            |
+--------------------------------------------------+-------------------------------------------------------------------------------------------------+
| logical_plan after datafusion                    | Projection: a.t1v1, a.t1v2, b.t1v1, b.t1v2                                                      |
|                                                  |   CrossJoin:                                                                                    |
|                                                  |     SubqueryAlias: a                                                                            |
|                                                  |       TableScan: t1                                                                             |
|                                                  |     SubqueryAlias: b                                                                            |
|                                                  |       TableScan: t1                                                                             |
| logical_plan after optd                          | LogicalProjection { exprs: [ #0, #1, #2, #3 ] }                                                 |
|                                                  | └── LogicalJoin { join_type: Cross, cond: true }                                                |
|                                                  |     ├── LogicalScan { table: t1 }                                                               |
|                                                  |     └── LogicalScan { table: t1 }                                                               |
| physical_plan after optd                         | PhysicalProjection { exprs: [ #0, #1, #2, #3 ] }                                                |
|                                                  | └── PhysicalNestedLoopJoin { join_type: Cross, cond: true }                                     |
|                                                  |     ├── PhysicalScan { table: t1 }                                                              |
|                                                  |     └── PhysicalScan { table: t1 }                                                              |
| physical_plan after optd-join-order              | (NLJ t1 t1)                                                                                     |
| physical_plan after optd-all-join-orders         | SAME TEXT AS ABOVE                                                                              |
| physical_plan after optd-all-logical-join-orders | (Join t1 t1)                                                                                    |
| physical_plan                                    | CollectorExec group_id=!17                                                                      |
|                                                  |   ProjectionExec: expr=[<expr>@0 as col0, <expr>@1 as col1, <expr>@2 as col2, <expr>@3 as col3] |
|                                                  |     CollectorExec group_id=!5                                                                   |
|                                                  |       CrossJoinExec                                                                             |
|                                                  |         CollectorExec group_id=!1                                                               |
|                                                  |           MemoryExec: partitions=1, partition_sizes=[0]                                         |
|                                                  |         CollectorExec group_id=!1                                                               |
|                                                  |           MemoryExec: partitions=1, partition_sizes=[0]                                         |
|                                                  |                                                                                                 |
+--------------------------------------------------+-------------------------------------------------------------------------------------------------+
```

### CODE REVIEW

File: optd-core/src/cascades/optimizer.rs [hunk 1]
```diff
@@ -209,13 +209,22 @@ impl<T: RelNodeTyp> CascadesOptimizer<T> {
...
+        let rel = self.memo.get_best_group_binding_with_meta(group_id, meta)?;
+        meta.insert(
+            rel.as_ref() as *const _ as usize,
+            RelNodeMeta::new(group_id),
+        );
+        Ok(rel)
+    }
+
+    /// Get the group binding.
+    pub fn step_get_optimize_rel(&self, group_id: GroupId) -> Result<RelNodeRef<T>> {
```
Comment (Gun9niR) [inline optd-core/src/cascades/optimizer.rs]: Do we still need this? Maintaining two versions of `step_get_optimizer_rel_[with_meta]` seems error-prone.

**Review by Gun9niR — COMMENTED** [review]
This feature is great! IIUC it also enables displaying logical properties or cost in explain stmt.

File: optd-core/src/cascades/optimizer.rs (hunk 1 above)
Comment (yliang412) [reply optd-core/src/cascades/optimizer.rs]: Yea maybe I will add an `Option<Meta>` and keep one copy.

Comment (Gun9niR) [reply optd-core/src/cascades/optimizer.rs]: When will this meta be none though?

Comment (yliang412) [reply optd-core/src/cascades/optimizer.rs]: You are right, I think I was mixing this one up with `memo.get_best_group_binding()`.

**Review by Gun9niR — APPROVED** [review]
lgtm

**Pull request merged**